### PR TITLE
JSON schemas for Agrifood datamodels

### PR DESCRIPTION
**Pull request by fdelavega**
JSON schemas for the agrifood datamodels defined at:
https://github.com/FIWARE/dataModels/blob/master/specs/AgriFood/README.md

### CODE REVIEW

**Comment by fdelavega** [comment]
Only one schema is being provided at this stage , I am creating the PR in order to allow you validate the  structure and avoid wasting time is not adequate

File: specs/AgriFood/AgriCrop/example.jsonld [hunk 1]
```diff
@@ -0,0 +1,76 @@
+{
+    "@context": [
+        "https://forge.etsi.org/gitlab/NGSI-LD/NGSI-LD/raw/master/coreContext/ngsi-ld-core-context.json",
+        "https://raw.githubusercontent.com/GSMADeveloper/NGSI-LD-Entities/master/examples/Agri-Crop-context.jsonld"
+    ],
+    "id": "urn:ngsi-ld:AgriCrop:df72dc57-1eb9-42a3-88a9-8647ecc954b4",
+    "type": "AgriCrop",
+    "createdAt": "2017-01-01T01:20:00Z",
+    "modifiedAt": "2017-05-04T12:30:00Z",
+    "source": "https://source.example.com",
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/example.jsonld]: source, dataProvider and entityVersion not yet adopted by ETSI, so we can remove them

File: specs/AgriFood/AgriCrop/example.jsonld [hunk 2]
```diff
@@ -0,0 +1,76 @@
+{
+    "@context": [
+        "https://forge.etsi.org/gitlab/NGSI-LD/NGSI-LD/raw/master/coreContext/ngsi-ld-core-context.json",
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/example.jsonld]: you can use 

http://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld

File: specs/AgriFood/AgriCrop/schema.json [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"id": {
+	    "type": "string",
+	    "format": "uri"
+	},
+	"type": {
+	    "type": "string",
+	    "enum": ["AgriCrop"],
+	    "description": "NGSI Entity Type"
+	},
+	"@context": {
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: no need to encode @context in the Schema

File: specs/AgriFood/AgriCrop/schema.json [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	    "$ref": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/common-schema.json#/definitions/PropertyString"
+	},
+	"agriSoil": {
+	    "$ref": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/common-schema.json#/definitions/Relationship"
+	},
+	"agriFertiliser": {
+	    "$ref": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/common-schema.json#/definitions/Relationship"
+	},
+	"agriPest": {
+	    "$ref": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/common-schema.json#/definitions/Relationship"
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: you are pointing to FiCodes ...

File: specs/AgriFood/AgriCrop/schema.json [hunk 5]
```diff
@@ -0,0 +1,110 @@
+{
+    "$schema": "http://json-schema.org/schema#",
+    "$id": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/AgriCrop/schema.json",
+    "title": "FIWARE - Agri Food",
+    "description": "This entity contains a harmonised description of a generic crop. This entity is primarily associated with the agricultural vertical and related IoT applications.",
+    "type": "object",
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: why don't you include common schemas from FIWARE Data Models? 

Please check other schemas

File: specs/AgriFood/AgriCrop/schema.json [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		    "type": "string",
+		    "enum": ["Property"]
+		},
+		"value": {
+		    "$ref": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/common-schema.json/definitions#/definitions/Interval"
+		}
+	    },
+	    "required": ["type", "value"]
+	},
+	"plantingFrom": {
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: at this level we don't need to encode all the NGSI-LD structure but just the property name and the value as in other data models

File: specs/AgriFood/common-schema.json [hunk 7]
```diff
@@ -0,0 +1,59 @@
+{
```
Comment (jmcanterafonseca) [inline specs/AgriFood/common-schema.json]: this is essentially not needed

File: specs/AgriFood/AgriCrop/example.jsonld [hunk 8]
```diff
@@ -0,0 +1,76 @@
+{
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/example.jsonld]: we would need examples as well for NGSIv2, both normalized and keyValues

File: specs/AgriFood/AgriCrop/schema.json (hunk 4 above)
Comment (samgh96) [reply specs/AgriFood/AgriCrop/schema.json]: I can't point to somewhere without data, I'll change this URL when I finish these schemas, in order to test them I need a valid URL.

File: specs/AgriFood/AgriCrop/example.jsonld (hunk 8 above)
Comment (samgh96) [reply specs/AgriFood/AgriCrop/example.jsonld]: Do I also need to implement a validator for NGSIv2?

File: specs/AgriFood/AgriCrop/example.jsonld (hunk 1 above)
Comment (samgh96) [reply specs/AgriFood/AgriCrop/example.jsonld]: Fixed in (https://github.com/Ficodes/dataModels/commit/871e5e062edb651fa2605a30b31f96b6cc0f5960)

File: specs/AgriFood/AgriCrop/example.jsonld (hunk 2 above)
Comment (samgh96) [reply specs/AgriFood/AgriCrop/example.jsonld]: Fixed in (https://github.com/Ficodes/dataModels/commit/23b9255ad972db1a60616447fd6099e57b12fa73)

File: specs/AgriFood/AgriCrop/schema.json (hunk 3 above)
Comment (samgh96) [reply specs/AgriFood/AgriCrop/schema.json]: Fixed in (https://github.com/Ficodes/dataModels/commit/9fc299b36222457f2e78536bc731cb9f0222258a)

File: specs/AgriFood/AgriCrop/schema.json (hunk 5 above)
Comment (samgh96) [reply specs/AgriFood/AgriCrop/schema.json]: Fixed in (https://github.com/Ficodes/dataModels/commit/b8969ba0ff3b70e9405c22ede6a35774a899d50d)

**Comment by jmcanterafonseca** [comment]
@samgh96 please follow the same approach as other schemas in this project i.e. do not try to validate against v2 or LD, just define the attributes, datatypes and values ... 

you can use this as reference

https://github.com/FIWARE/dataModels/blob/master/specs/WasteManagement/WasteContainer/schema.json 

/cc @fdelavega

**Comment by jmcanterafonseca** [comment]
be careful with the examples:

you need

example.json (Simplified format. Do not reference any @context)
example-normalized.json (Normalized format NGSIv2)
example-normalized-ld.jsonld (Normalized format NGSI-LD)

And also you would need to modify the validator so that the .jsonld files are not tested against the Context Broker, as currently they will fail (or add additional tests against Orion-LD, as you prefer).

File: specs/AgriFood/AgriCrop/example-normalized.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
+{
+    "id": "urn:ngsi-ld:AgriCrop:df72dc57-1eb9-42a3-88a9-8647ecc954b4",
+    "type": "AgriCrop",
+    "createdAt": {
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/example-normalized.json]: not needed in v2, in v2 these attributes are named dateCreated, dateModified 

BTW be careful the NGSI type is "DateTime" "date-time" is a JSON Schema type

File: specs/AgriFood/AgriCrop/schema.json [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                    "type": "string"
+                },
+                "alternateName": {
+                    "type": "string"
+                },
+                "description": {
+                    "type": "string"
+                },
+                "agriSoil": {
+                    "type": "string",
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: schema is wrong, this should be an array

File: specs/AgriFood/AgriCrop/schema.json [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                    "format": "uri"
+                },
+                "agriFertiliser": {
+                    "type": "string",
+                    "format": "uri"
+                },
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: schema is wrong this should be an array

File: specs/AgriFood/AgriCrop/schema.json [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                "agriPest": {
+                    "type": "string",
+                    "format": "uri"
+                },
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: schema is wrong this should be an Array

File: specs/AgriFood/AgriCrop/schema.json [hunk 13]
```diff
@@ -0,0 +1,94 @@
+{
+    "$schema": "http://json-schema.org/schema#",
+    "$id": "https://raw.githubusercontent.com/Ficodes/dataModels/agriDataModels/specs/AgriFood/AgriCrop/schema.json",
+    "title": "FIWARE - Agri Food",
+    "description": "This entity contains a harmonised description of a generic crop. This entity is primarily associated with the agricultural vertical and related IoT applications.",
+    "type": "object",
+    "allOf": [
+        { "$ref": "https://fiware.github.io/dataModels/common-schema.json#/definitions/GSMA-Commons" },
+        { "$ref": "https://fiware.github.io/dataModels/common-schema.json#/definitions/Location-Commons" },
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: do we really need this schema?

File: specs/AgriFood/AgriCrop/schema.json [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        { "$ref": "https://fiware.github.io/dataModels/common-schema.json#/definitions/PhysicalObject-Commons" },
```
Comment (jmcanterafonseca) [inline specs/AgriFood/AgriCrop/schema.json]: do we really need this schema?

**Pull request merged**